src/pybind/mgr/cephadm/services/monitoring.py

```diff
@@ -256,86 +255,46 @@ def generate_config(
             self,
             daemon_spec: CephadmDaemonDeploySpec,
     ) -> Tuple[Dict[str, Any], List[str]]:
+
         assert self.TYPE == daemon_spec.daemon_type
         deps = []  # type: List[str]
-
-        # scrape mgrs
-        mgr_scrape_list = []
-        mgr_map = self.mgr.get('mgr_map')
         port = cast(int, self.mgr.get_module_option_ex(
             'prometheus', 'server_port', self.DEFAULT_MGR_PROMETHEUS_PORT))
         deps.append(str(port))
-        t = mgr_map.get('services', {}).get('prometheus', None)
+        # add an explicit dependency on the active manager. This will force to
+        # re-deploy prometheus if the mgr has changed (due to a fail-over i.e).
+        deps.append(self.mgr.get_active_mgr().name())
+
+        t = self.mgr.get('mgr_map').get('services', {}).get('prometheus', None)
+        sd_port = self.mgr.service_discovery_port
+        srv_end_point = ''
         if t:
             p_result = urlparse(t)
             # urlparse .hostname removes '[]' from the hostname in case
             # of ipv6 addresses so if this is the case then we just
             # append the brackets when building the final scrape endpoint
             if '[' in p_result.netloc and ']' in p_result.netloc:
-                mgr_scrape_list.append(f"[{p_result.hostname}]:{port}")
+                srv_end_point = f'https://[{p_result.hostname}]:{sd_port}/sd/prometheus/sd-config?'
```

Comment:
We can do it either way. My plan was to move to the new service discovery and make it the default configuration. The only thing that I'd make configurable is to enable/disable the up-coming security feature. WDYT?

Reply:
I tend to like leaving old, working methods as an option when we overhaul things in case the new way turns out to be broken. I guess it depends how confident we are this is going to fully work.

Reply:
By now the idea is to go with the new service-discovery based implementation and deprecate the old static conf but that would depend mainly on the stability of this new impl.